voxblox_ros/src/simulation_server.cc

```diff
@@ -0,0 +1,417 @@
+
+    // Put into the real map.
+    tsdf_integrator_->integratePointCloud(T_G_C, ptcloud_C, colors);
+
+    if (generate_occupancy_) {
+      occ_integrator_->integratePointCloud(T_G_C, ptcloud_C);
+    }
+
+    const bool clear_updated_flag = true;
+    if (incremental_) {
+      esdf_integrator_->updateFromTsdfLayer(clear_updated_flag);
+    }
+
+    // Convert to a XYZRGB pointcloud.
+    if (visualize_) {
+      ptcloud_pcl.header.frame_id = world_frame_;
+      pcl::PointXYZRGB point;
+      point.x = view_origin.x();
+      point.y = view_origin.y();
+      point.z = view_origin.z();
+      ptcloud_pcl.push_back(point);
+
+      view_ptcloud_pub_.publish(ptcloud_pcl);
+      ros::spinOnce();
+    }
+  }
+
+  // Generate ESDF in batch.
+  if (!incremental_) {
+    if (generate_occupancy_) {
+      esdf_occ_integrator_->updateFromOccLayerBatch();
+    }
+
+    esdf_integrator_->updateFromTsdfLayerBatch();
+
+    // Other batch options for reference:
+    // esdf_integrator_->updateFromTsdfLayerBatchFullEuclidean();
+    // esdf_integrator_->updateFromTsdfLayerBatchOccupancy();
+  }
+}
+
+template <typename VoxelType>
+FloatingPoint SimulationServer::evaluateLayerAgainstGt(
+    const Layer<VoxelType>& layer_test,
+    const Layer<VoxelType>& layer_gt) const {
+  // Iterate over all voxels in GT set and look them up in the test set.
+  // Then compute RMSE.
+  BlockIndexList block_list;
+  layer_gt.getAllAllocatedBlocks(&block_list);
+  size_t vps = layer_gt.voxels_per_side();
+  size_t num_voxels_per_block = vps * vps * vps;
+
+  double total_squared_error = 0.0;
+  size_t num_evaluated_voxels = 0;
+
+  for (const BlockIndex block_index : block_list) {
+    if (!layer_test.hasBlock(block_index)) {
+      continue;
+    }
+    const Block<VoxelType>& gt_block = layer_gt.getBlockByIndex(block_index);
+    const Block<VoxelType>& test_block =
+        layer_test.getBlockByIndex(block_index);
+
+    for (size_t linear_index = 0; linear_index < num_voxels_per_block;
+         ++linear_index) {
+      FloatingPoint error = 0.0;
+      if (evaluateVoxel<VoxelType>(
+              test_block.getVoxelByLinearIndex(linear_index),
+              gt_block.getVoxelByLinearIndex(linear_index), &error)) {
+        total_squared_error += error * error;
+        num_evaluated_voxels++;
+      }
+    }
+  }
+  // Return the RMSE.
+  if (num_evaluated_voxels == 0) {
+    return 0.0;
+  }
+  ROS_INFO_STREAM("Number of voxels evaluated: " << num_evaluated_voxels);
+  return sqrt(total_squared_error / num_evaluated_voxels);
+}
+
+template <>
+bool SimulationServer::evaluateVoxel(const TsdfVoxel& voxel_test,
+                                     const TsdfVoxel& voxel_gt,
+                                     FloatingPoint* error) const {
+  if (voxel_test.weight < 1e-6 || voxel_gt.distance < 0.0) {
+    return false;
+  }
+
+  *error = voxel_gt.distance - voxel_test.distance;
+
+  /* if (voxel_gt.distance < -truncation_distance_) {
```

Comment:
why is this commented out?

Reply:
Ahh I think this should be deleted -- this has to do with how behind-surface measurements are penalized in the evaluation, but I eventually settled on only evaluating positive distances (I think this is mentioned in the paper) and the if statement above confirms this. So yeah should just be deleted, here and below.